✨ [Feature] 통계 API 구현

jis-kim opened the pull request:
## Summary 
- 다섯 개의 쿼리에 따른 posting 의 통계를 제공하는 `GET /posting` API 를 구현했습니다.
- `hashtag`, `type`, `start`, `end`, `value`

## Describe your changes
- `StatisticsModule`, `StatisticsService`, `StatisticsController` 파일 추가.
- `[start,end]` 기간에 `hashtag`가 쓰여 작성된  `Posting`의 갯수, 또는 `value`의 합계를 `type`에 따라 가져오는 통계 API 구현.
  - 기본적인 query validation 은 `StatisticQueryDto` 에서 `class-validator`의 데코레이터를 이용했습니다.
  - `start`, `end` 체크는 복잡한 validation 과정이 필요했으므로 `service` 레이어에서 `validateDate` 함수를 이용하여 처리했습니다.
  - query parameter 의 기본 값들은 타입을 좀 더 명확히 정의하기 위해 `setDefaultValue` 함수를 이용하여 세팅해주었습니다. `StatistricQueryDto`의 property 가 optional 하기 때문에 로직상 `undefined` 인 property가 없지만 `undefined` 타입 에러가 뜨는 문제를 해당 함수 return 값을 `Required` utility 타입을 설정하여 해결했습니다.
- ESLint 규칙을 추가했습니다. (`**/*.constants.ts`, `**/*.decorator.ts`)

## 추가 사항

- [x] - 도연님 PR #20 머지 후 `userName` 가져오는 로직 추가하겠습니다.

## Issue number and link
close #14

jis-kim committed:
:hammer: chore: validation에 필요한 패키지 설치

jis-kim committed:
:sparkles: feat: statisticQuerDto 정의 #14

- class-validator 를 이용한 기본 검증 로직 구현

jis-kim committed:
:hammer: chore: 불필요한 app module 파일 삭제

jis-kim committed:
:sparkles: feat: 쿼리 검증을 위한 서비스 함수 구현 #14

- start, end 유효성 검증하는 validateDate
- 'date' type일 때 시간을 00시 ~ 23시59분59초999로 고정하는 adjustDate

jis-kim committed:
:sparkles: feat: 쿼리 파라미터에 따라 데이터 가져오는 서비스 함수 구현 #14

- date, hour에 따라 다른 형식으로 return

jis-kim committed:
:lipstick: style: lint 규칙에 맞게 함수 return type 추가

jis-kim commented on src/statistics/statistics.service.ts:
```diff
@@ -0,0 +1,118 @@
+   * @param end
+   */
+  private adjustDate(start: Date, end: Date): void {
+    //NOTE: local time (KST) 기준으로 실행되고 있으므로 논의가 필요.
```
KST 기준으로 진행해도 문제없다고 판단되어 resolve 합니다.

kimdoyeonn commented on src/statistics/statistics.service.ts:
```diff
@@ -0,0 +1,118 @@
+      this.adjustDate(start, end);
+    }
+
+    // 현 시간보다 이후의 시간이거나, start가 end보다 클 경우 false 반환
```
시작 시간이 끝나는 시간보다 늦을 경우엔 에러를 보내도록 하는게 더 좋을것 같습니다.

jis-kim commented on src/statistics/statistics.service.ts:
```diff
@@ -0,0 +1,118 @@
+    return this.dataSource
+      .getRepository(Posting)
+      .createQueryBuilder('posting')
+      .select(value === 'count' ? 'COUNT(posting.id)' : `SUM(posting.${value})`, 'value')
```
따로 const 사용하는 것으로 수정

jis-kim committed:
:hammer: chore: 상수값, 데코레이터 이름 규칙을 위한  ESLint 설정 수정

- 상수는 UPPER_CASE, 데코레이터는 PascalCase 로 적용하기 위해 규칙 수정
- decorator, constants 파일을 lint 에서 제외합니다.

jis-kim committed:
:lipstick: style: 바뀐 Eslint 규칙에 따라 상수파일 이름을 .contants.ts 파일로  변경

jis-kim committed:
:lipstick: style: 주석 추가 및 형식 변경

- service 함수 parameter, return type 주석 추가

jis-kim committed:
:sparkles: feat: start, end date validation 수정

- type이 'date' 일 때 start<end 여도 validation 통과하는 문제를 수정했습니다.

jis-kim committed:
:recycle: refactor: statistic query dto 를 명확히 수정하고 기본값 설정을 service로 이동

- 기존 dto 에서 기본값 설정하던 것을 service로 옮기고 dto optional 값을 optional type으로 설정.
- response 타입 에러는 Required  유틸 타입으로 해결.

jis-kim committed:
Merge branch 'dev' of github.com:INgandle/social-feed-hub into feature/statistics

jis-kim committed:
Merge branch 'dev' of github.com:INgandle/social-feed-hub into feature/statistics

jis-kim committed:
:sparkles: feat: auth 로직 적용하여 getStatistic 에서 userName을 가져오도록 수정  #14

- req.user의 userName을 가져와서 실제 api 동작에 반영했습니다.

ooheunda commented on src/main.ts:
```diff
@@ -12,6 +13,13 @@ const config = new DocumentBuilder()
 
+  app.useGlobalPipes(
+    new ValidationPipe({
+      stopAtFirstError: true, // 맨 처음 발생한 하나의 에러만 반환 (나머지 검증 skip)
```
ValidationPipe에 이 옵션은 처음 보는데 너무 좋은 것 같아요! 진작에 알았으면 좋았을걸...

jis-kim replied:
이거 근데 ...제가 생각한대로 안되더라고요......... 저는 dto내에서 각 프로퍼티 마다 적용되는 줄 알았는데 검증할 객체가 여러 개인 케이스에 적용되는 옵션같습니다...

ooheunda replied:
아?! ㅋㅋㅋㅋ 저도 그렇게 이해하고 좋다 생각했는데 테스트해보니 정말이네여... 아쉽다...

ooheunda commented on src/statistics/statistics.service.ts:
```diff
@@ -0,0 +1,135 @@
+      .createQueryBuilder('posting')
+      .select(aggregationQuery, 'value')
+      .addSelect(`DATE_FORMAT(posting.createdAt, ${dateFormat})`, 'date')
+      .innerJoin('posting.postingHashtags', 'postingHashtag')
+      .innerJoin('postingHashtag.hashtag', 'hashtag')
```
이땐 where 조건 자체가 `hashtag`라서, 무조건 `hashtag`가 있는 레코드들만 조인하게 되니 leftJoin이 아니라 innerJoin을 해도 되는거죠?!
아까 말씀해주신 부분이긴 한데 알맞게 이해한건지 확인용으로 달아봅니다ㅎㅎ...

jis-kim replied:
네! 해당 쿼리는 "posting에 사용된 hashtag 중에" hashtag.name이 name이고 start end 사이에 있는 레코드를 집계하는데요, "" 사이에 쓰인 부분이 join이라고 보시면 됩니다. hashtag, posting 둘 중 하나의 레코드가 없는 데이터는 필요없어서 `innerJoin`을 사용했습니다.
`hashtag.name`조건으로 셀렉하므로 `leftJoin`을 해도 똑같은 결과가 나오겠지만 만약 `leftJoin`을 사용한다면 join 상태의 테이블이 커져서 where 에 더 큰 비용이 발생할 듯 합니다!

ooheunda reviewed (approved):
고생하셨습니다~👍
코멘트는 별 거 아니라 Approve 합니닷

kimdoyeonn reviewed (approved):
수고하셨습니다~
